pkg/sensors/accelerator/gpu/source/gpu_dcgm.go

```diff
@@ -0,0 +1,326 @@
+	}
+	d.collectionSupported = false
+	return true
+}
+
+func (d *GPUDcgm) GetAbsEnergyFromGPU() []uint32 {
+	gpuEnergy := []uint32{}
+	for _, device := range d.devices {
+		power, ret := device.(nvml.Device).GetPowerUsage()
+		if ret != nvml.SUCCESS {
+			klog.V(2).Infof("failed to get power usage on device %v: %v\n", device, nvml.ErrorString(ret))
+			continue
+		}
+		// since Kepler collects metrics at intervals of SamplePeriodSec, which is greater than 1 second, it is
+		// necessary to calculate the energy consumption for the entire waiting period
+		energy := uint32(uint64(power) * config.SamplePeriodSec)
+		gpuEnergy = append(gpuEnergy, energy)
+	}
+	return gpuEnergy
+}
+
+func (d *GPUDcgm) GetGpus() map[string]interface{} {
+	return d.devices
+}
+
+func (d *GPUDcgm) GetProcessResourceUtilizationPerDevice(device interface{}, deviceName string, since time.Duration) (map[uint32]ProcessUtilizationSample, error) {
+	processAcceleratorMetrics := map[uint32]ProcessUtilizationSample{}
+
+	if device == nil { // this is a MIG device, it is already tracked in the parent device
+		return processAcceleratorMetrics, nil
+	}
+
+	var vals, miVals []dcgm.FieldValue_v1
+	var err error
+
+	klog.Infof("Device %v\n", deviceName)
+	if device != nil { // this is a GPU
+		deviceIndex, strErr := strconv.Atoi(deviceName)
+		if strErr != nil {
+			klog.Infof("failed to convert %q to an integer: %v", deviceName, strErr)
+			return processAcceleratorMetrics, strErr
+		}
+		vals, err = dcgm.GetLatestValuesForFields(uint(deviceIndex), deviceFields)
+		if err != nil {
+			klog.Infof("failed to get latest values for fields: %v", err)
+			return processAcceleratorMetrics, err
+		}
+		if err == nil {
+			for i, val := range vals {
+				label := deviceFieldsString[i]
+				klog.Infof("Device %v Label %v Val: %v", deviceName, label, ToString(val))
+			}
+			klog.Infof("\n")
+		}
+		processInfo, ret := device.(nvml.Device).GetComputeRunningProcesses()
+		if ret != nvml.SUCCESS {
+			klog.Infof("failed to get running processes: %v", nvml.ErrorString(ret))
+			return processAcceleratorMetrics, fmt.Errorf("failed to get running processes: %v", nvml.ErrorString(ret))
+		}
+		for _, p := range processInfo {
+			// klog.Infof("pid: %d, memUtil: %d gpu instance id %d compute id %d\n", p.Pid, p.UsedGpuMemory, p.GpuInstanceId, p.ComputeInstanceId)
+			if p.GpuInstanceId > 0 { // this is a MIG, get it entity id and reads the related fields
+				entityName := gpuMigArray[deviceIndex][p.GpuInstanceId].EntityName
+				multiprocessorCountRatio := gpuMigArray[deviceIndex][p.GpuInstanceId].MultiprocessorCountRatio
+				mi := d.entities[entityName]
+				miVals, err = dcgm.EntityGetLatestValues(mi.EntityGroupId, mi.EntityId, deviceFields)
+				if err == nil {
+					for i, val := range miVals {
+						label := deviceFieldsString[i]
+						value := ToString(val)
+						klog.Infof("Device %v Label %v Val: %v", entityName, label, value)
+						if val.FieldId == dcgm.DCGM_FI_PROF_SM_ACTIVE {
```

Comment:
we can use `DCGM_FI_PROF_PIPE_TENSOR_ACTIVE` instead of `DCGM_FI_PROF_SM_ACTIVE` here, since `DCGM_FI_PROF_PIPE_TENSOR_ACTIVE` looks closer to actual  resource utilization rate

Reply:
updated